### DISCUSSION TITLE
How Does Nullsafety in C# Fit Into the Type System?

### DISCUSSION THREAD

**Discussion by moneer-muntazah** [dotnet/csharplang#6471] [Q&A] [answered]
Hi

I come from Dart background I want to have a deeper understanding of how null safety works in C#. In Dart, it is easy for me to understand null safety in terms of type hierarchy for example `String` and `Null` are subtypes of `String?` and the `!` operator performs a cast operation. How is it different in C# was there a change to the .NET data types hierarchy? When I use the unary `!` I'm I casting types? Is the `?` operator just an annotation? Or does it state a different type altogether? 

Thank you all. I went through the docs, but couldn't get the right answers.

**Comment by En3Tho** [comment, marked as answer]
It's just a compile-time annotation to help developers be wary of potential nulls. Under the hood string or string? is the same type. "!" is just a "a know better than you, compiler" indicator. There are no casts involved.

It's a different story with Nullable<T> type - it's designed to enable "nullness" for ValueTypes. It's essentially a wrapper - a different type - which holds some value inside it and indicates whether that value was initialized properly or not (kinda "null"). It has a special runtime treatment in some cases like boxing.

So basically "referenceType? (e.g. string?)" is just an compile time annotation and "valueType? (e.g. int?)" is a container type wrapping original value.

**Reply by TahirAhmadov** [reply]
Also, the `!` operator can also be used with `int?` - the compiler will warn if you do `int? x = null; int y = x.Value;`, and you can suppress it with `int y = x!.Value;`
